### ISSUE TITLE
[DependencyInjection] Slow performance due to case insensitivity inside resolveEnvPlaceholders

### ISSUE THREAD

**Issue by nicoschoenmaker** [symfony/symfony#28292]
**Symfony version(s) affected**: 3.2 and up

**Description**  
Running `bin/console` with an empty cache spends 30.5% of the time inside `resolveEnvPlaceholders`, which is due to case insensitive string matching.

**How to reproduce**
Running `bin/console` in an application that has been converted to use (lots of) environment variables. Then make some coffee :sleeping:.

The application I'm working on connects to many different suppliers, and has 247 (!) environment variables because of that.

**Possible Solution**
A possible solution would be to make this case sensitive, since that is [faster](https://stackoverflow.com/a/33354124). While this did not change behaviour in my application, this might be a BC break?

Applying this patch made the `bin/console` run **~6% faster** for me.
```diff
--- ContainerBuilder.php	2018-08-28 09:10:05.873657614 +0200
+++ ContainerBuilder.php	2018-08-28 09:16:58.721668989 +0200
@@ -1366,7 +1366,7 @@
         $completed = false;
         foreach ($envPlaceholders as $env => $placeholders) {
             foreach ($placeholders as $placeholder) {
-                if (false !== stripos($value, $placeholder)) {
+                if (false !== strpos($value, $placeholder)) {
                     if (true === $format) {
                         $resolved = $bag->escapeValue($this->getEnv($env));
                     } else {
@@ -1379,7 +1379,7 @@
                         if (!\is_string($resolved) && !is_numeric($resolved)) {
                             throw new RuntimeException(sprintf('A string value must be composed of strings and/or numbers, but found parameter "env(%s)" of type %s inside string value "%s".', $env, \gettype($resolved), $this->resolveEnvPlaceholders($value)));
                         }
-                        $value = str_ireplace($placeholder, $resolved, $value);
+                        $value = str_replace($placeholder, $resolved, $value);
                     }
                     $usedEnvs[$env] = $env;
                     $this->envCounters[$env] = isset($this->envCounters[$env]) ? 1 + $this->envCounters[$env] : 1;
```

**Additional context**  
This is a screenshot of kcachegrind of a `bin/console` run with an empty `var` directory. It denotes that 30.5% of the time is spent inside `resolveEnvPlaceholders`, with 62% of that due to `stripos` and 31% due to `str_ireplace`.

![screenshot from 2018-08-28 09 21 59](https://user-images.githubusercontent.com/1469323/44707131-f0bb5480-aaa3-11e8-8ef8-cbdeb036b6e5.png)

Measurements in seconds

| Case insensitive (current) | Case sensitive |
|----------------------------|----------------|
|  31.3                      | 29.3           |
| 30.6                       | 28.7           |
| 30.9                       | 28.7           |
| 30.9                       | 30             |
| 31.5                       | 29             |
| Average **31.04**                      | Average **29.14**          |

Would be willing to work towards a PR for this issue, but I'm mostly looking how to speed up this piece of code without a BC break :)

**Comment by nicolas-grekas** [comment]
An idea to fix this would be to make placeholders be only numbers.
See this line: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/ParameterBag/EnvPlaceholderParameterBag.php#L57
The prefix has to contain letter, but we search for the number-only part before using stripos.
PR welcome on master.

**Comment by nicolas-grekas** [comment]
BTW, using https://blackfire.io/ to do profiles is so much better ;)

**Comment by stof** [comment]
why are we making env placeholders case insensitive anyway ? such placeholders are opaque to the code receiving them, so allowing them to change the case looks weird.

**Comment by nicolas-grekas** [comment]
The reason is that di extension do process config values (e.g. using strtolower/strtoupper), and we still want to pass through these.

**Comment by javiereguiluz** [comment]
@nicoschoenmaker thanks for your report and all the details!

My guess is that most apps have a few env vars, but this app has 247 !! Also, the improvement doesn't look perceptible to end users: waiting for 31.04 seconds or 29.14 seconds will be awfully long in both cases.

I'd say this is an edge case and we shouldn't change it unless we receive similar reports from other apps. But let's hear more opinions! Thanks!

**Comment by nicolas-grekas** [comment]
I agree with Javier. I'd even add that 247 env card is suspicious: do you really have that many vars to adjust your app for your infrastructure? Maybe you need to use more parameters instead, for everything that is not infrastructure related.

**Comment by nicoschoenmaker** [comment]
The amount of environment variables surprised me, but we need just about all of them. We would be able to get rid of about 30-40 if we'd really try, but we'd still end up with >200 of them.

Getting it down to 29 seconds is not my goal no, it should be much faster than that. But this was the first thing that I spotted that had a measurable impact. Hence only described as possible solution :slightly_smiling_face:

**Comment by nicolas-grekas** [comment]
@nicoschoenmaker if you're up to working on a patch to improve this, please do (start with [the idea proposed above](https://github.com/symfony/symfony/issues/28292#issuecomment-416623627) maybe). Otherwise, let's close IMHO.

**Comment by svandervlugt** [comment]
I've been looking at a different approach, which gained me a 25% speed increase on bin/console for our project.
I noticed the ```doExport``` function in the PhpDumper being called with the same variables a lot.

https://github.com/svandervlugt/symfony/commit/100adda27fedbf1708ba66ff16fd9cf3a92b498a

**Comment by nicolas-grekas** [comment]
25% of what? You're going from 30s to 23s with this change?

**Comment by svandervlugt** [comment]
We also did some optimizations on the project side, lowering the numbers in general.
These are the results of the change linked in https://github.com/symfony/symfony/issues/28292#issuecomment-423914537:

| Before     | After      |
|------------|------------|
| 24.1       | 15.8       |
| 24.5       | 16.0       |
| 24.7       | 16.3       |
| 24.2       | 16.0       |
|  Avg: 24.4 | Avg:16.0   |

Closer to a 33% speed increase instead of the earlier mentioned 25%.
